### ISSUE TITLE
Optimize genesis file chunking

### ISSUE THREAD

**Issue by andynog** [cometbft/cometbft#1289]
As part of #1037, we need to optimize how the genesis file is chunked.

we need to implement logic to:

- Determine the number of chunks (based on the chunk size and file size). This logic already exists, might need to adapt for optimal chunk size.
- Store the chunks in the file system (TBD in the `tmp` or `data` folder inside home)
- Return chunks loaded from the file system (not from storage) for the `genesis_chunked` endpoint.
- [OPTIONAL] Validate chunks (re-create the file from chunks) and hash it to ensure it matches hash in storage or hash of genesis in the file system (`config` folder). Maybe this could be validated when the node starts to ensure nothing has been tampered with.

#### DOD 
- [x] Do not store chunked genesis files in memory. 
- [x] Store the chunks on storage or file system and fetch them on demand. 
- [x] Rethink the default size of the genesis chunk and use an appropriate value.

This completes the work started in Phase II of #1037, thus finalizing the implementation of the chunking strategy discussed [here](https://github.com/cometbft/cometbft/issues/1289#issuecomment-2422499013).

**Comment by jmalicevic** [comment]
Can we avoid storing the file twice by using file pointers to point into the file content based on chunk boundaries and then return those?

**Comment by alesforz** [comment]
After discussions in [#1037](https://github.com/cometbft/cometbft/issues/1037), [#1286](https://github.com/cometbft/cometbft/issues/1286), [#1290](https://github.com/cometbft/cometbft/issues/1290), and offline, we have some thought about how to proceed.

Firstly, we require chunking only to support the `/genesis_chunked` JSON-RPC endpoint. Please refer to [this comprehensive comment](https://github.com/cometbft/cometbft/issues/1290#issuecomment-2377326927) for our plans to tackle the 'chunking problem' when serving JSON-RPC requests.

The gRPC APIs require a different strategy. We do not plan to support a gRPC 'chunking' endpoint. Instead, the gRPC `/genesis` endpoint will always return the genesis file using gRPC streaming. For pseudocode showing how this endpoint could be implemented, please see [this comment](https://github.com/cometbft/cometbft/issues/1037#issuecomment-2361424581).

Given this, to serve the gRPC APIs we could store the genesis file only on disk. The main challenge is that for each request, we need to re-hash the genesis file and compare its hash against the one stored in our database to ensure it hasn't been tampered with. We can perform this _without_ loading the entire file into memory by using the `io.Copy` function like so:
```go
// source is pointing to the genesis file on disk.
func SumStream(source io.Reader) ([]byte, error){
	hash := sha256.New()

	if _, err := io.Copy(hash, source); err != nil {
		// handle error
	}
	return hash.Sum(nil), nil
}
```
This works because the `io.Copy` function streams data from an `io.Reader` to an `io.Writer`, reading the source in chunks and writing those chunks to the destination without loading the entire source into memory at once. Similarly, `sha256.New()` returns a hash that can be updated incrementally with each chunk of data. Therefore, we can hash the file as we read it, eliminating the need to load the entire file into memory for the hashing process.

This method allows us to avoid storing the genesis file in memory. However, there are trade-offs:
- **Re-hashing Overhead**: For each request, we have to re-hash the genesis file. For large genesis files (potentially several GBs), this introduces significant overhead.
- **Double File Reading**: The genesis file is read twice per request—once to compute the hash and once to stream it to the client. While we could hash and stream the file in a single pass, if the computed hash doesn't match the one in our database, the client would have already received an invalid genesis file by the time we detect the mismatch.

If these performance trade-offs are acceptable, we can avoid storing in-memory copies. If not, we may need to store the genesis file in memory (perhaps in chunks, as we plan to do for the JSON-RPC APIs), allowing the streaming API to return the data directly without reading from disk each time.

(cc @jmalicevic @andynog)

**Comment by sergio-mena** [comment]
Hi @alesforz, sorry for delay reacting to [this](https://github.com/cometbft/cometbft/issues/1289#issuecomment-2377450211). Here are my comments.

The hashing mechanism looks excellent to me (removing the need to load the genesis file into memory). But the question I ask myself is:

* What is the model we are considering here, in terms of Byzantine behaviour?
* I always had the impression that we assumed correct nodes (i.e., non-Byzantine nodes) are _not_ compromised in any way. So their file system can be trusted... at least while the node is running.
* If we agree to be under that model, I don't think we need to check the hash of the genesis file _every time_ a gRPC client requests it (even without loading it into memory, a big genesis file will require a sizeable amount of CPU and I/O resources).
* Just as a sanity check, even if not strictly required by the model I mentioned above, maybe we can check the hash (with the efficient `io.Copy` usage you present above) only once:
  * either at start up
  * or upon the first request for the genesis file from a gRPC client since the node last restarted
 
This last point covers the danger that, as our system is crash-recovery, the file system is modified during downtime (which might last months, as we recently saw in Polygon testnets). However, I strongly believe that we should keep assuming that the file system can be trusted while the node is running... otherwise, why trust e.g. the blockstore DB, which is the source of the `/block` RPC and is stored in the same file system?.

**Comment by alesforz** [comment]
Thanks @sergio-mena!

If we work under the assumption that correct nodes aren't compromised, re-hashing the genesis or its chunks is unnecessary. We will only retrieve the genesis/chunk from disk and send it to the requesting client(s). 

The `Node` already does a sanity check of the genesis file at startup, so we don't need to add code for that.

**Comment by alesforz** [comment]
Here is the updated strategy:
- **Node Type Modification**: The [`Node`](https://github.com/cometbft/cometbft/blob/2c8d1e4909e9ee287428869fa233f631a95c4f67/node/node.go#L50) type will no longer store a `GenesisDoc`. As explained in [this comment](https://github.com/cometbft/cometbft/issues/1286#issuecomment-2353309745), the `Node` only requires the `GenesisDoc.GenesisTime` field after initialization. Therefore, the `Node` will now store the `GenesisTime` directly, eliminating the need to hold a pointer to the entire genesis document, thus removing one copy of the genesis from memory.
- **Environment Type Changes**: The [`Environment`](https://github.com/cometbft/cometbft/blob/9e25845d45329ad17dddbace0345e8e8514926c5/rpc/core/env.go#L72) type will store the genesis file's path on disk. If the genesis file exceeds a pre-configured size threshold, it will split it into chunks that it will store on disk. 
- **API Compatibility**: These changes to the `Environment` do *not* break the existing JSON-RPC APIs. Users of `/genesis` and `/genesis_chunked` can continue using them as before. 

#### How will the `/genesis_chunked` work under the hood?
Upon the reception of a request for a chunk, it will retrieve the chunk from disk and return it.

#### What are the advantages of this solution?
With the proposed solution, we won't keep the genesis in memory anymore. Given that some chains have genesis files that are GBs in size, this is a subtantial memory save.

#### Tradeoffs
Currently, the JSON-RPC APIs are very responsive because the implementation stores the genesis file and its chunks in memory. Storing the chunks on disk will increase the time required to serve each request because of disk I/O. We believe this is a reasonable trade-off given the runtime memory savings this proposal offers.

#### gRPC APIs
We are currently [implementing](https://github.com/cometbft/cometbft/issues/2693#3633) the gRPC APIs, which will support downloading the genesis file only via gRPC streaming. Therefore, we will no longer need to store chunks of the genesis file for these APIs.

**Comment by alesforz** [comment]
Here is a visualization of the changes, to be compared with the [previous wave of improvements](https://github.com/cometbft/cometbft/issues/1290#issuecomment-2417082937):

![profile.pdf](https://github.com/user-attachments/files/17555662/profile.pdf)

With a genesis file of ~325 MB the changes implemented in this issue bring a `Node` memory usage down to about 1.5MB. That is, these changes result in a 99% reduction in memory usage.
